contracts/src/zaps/UniV3Zap.sol

```diff
@@ -0,0 +1,937 @@
+        // properly sized for the proceeds that need to be converted.
+        //
+        // NOTE: Use the zap contract's balance rather than the proceeds
+        // reported by Hyperdrive to avoid having to handle the difference
+        // between rebasing and non-rebasing tokens.
+        uint256 balance = IERC20(tokenIn).balanceOf(address(this));
+        _swapParams.amountOutMinimum = balance.mulDivDown(
+            _swapParams.amountOutMinimum,
+            _swapParams.amountIn
+        );
+        _swapParams.amountIn = balance;
+
+        // If the output token is WETH, we always unwrap to ETH. In this case,
+        // we make this contract the recipient of the Uniswap swap.
+        address tokenOut = _swapParams.path.tokenOut();
+        address recipient = _swapParams.recipient;
+        if (tokenOut == address(weth)) {
+            _swapParams.recipient = address(this);
+        }
+
+        // Execute the Uniswap trade.
+        //
+        // NOTE: We increase the required approval amount by 1 wei so that the
+        // router ends with an approval of 1 wei. This makes future approvals
+        // cheaper by keeping the storage slot warm.
+        ERC20(tokenIn).forceApprove(
+            address(swapRouter),
+            _swapParams.amountIn + 1
+        );
+        proceeds = swapRouter.exactInput(_swapParams);
+
+        // If the output token is WETH, unwrap the WETH and send the ETH.
+        if (tokenOut == address(weth)) {
+            weth.withdraw(weth.balanceOf(address(this)));
+            (bool success, ) = recipient.call{ value: address(this).balance }(
+                ""
+            );
+            if (!success) {
+                revert TransferFailed();
+            }
+        }
+
+        return proceeds;
+    }
+
+    /// @dev Converts a quantity in base to vault shares. This works for all
+    ///      Hyperdrive pools.
+    /// @param _hyperdrive The Hyperdrive instance.
+    /// @param _baseAmount The base amount.
+    /// @return The converted vault shares amount.
+    function _convertToShares(
+        IHyperdrive _hyperdrive,
+        uint256 _baseAmount
+    ) internal view returns (uint256) {
+        // If this is a mainnet deployment and the address is the legacy stETH
+        // pool, we have to convert the proceeds to shares manually using Lido's
+        // `getSharesByPooledEth` function.
+        if (
+            block.chainid == 1 &&
+            address(_hyperdrive) == LEGACY_STETH_HYPERDRIVE
+        ) {
+            return
+                ILido(_hyperdrive.vaultSharesToken()).getSharesByPooledEth(
+                    _baseAmount
+                );
+        }
+        // If this is a mainnet deployment and the address is the legacy stETH
+        // pool, we have to convert the proceeds to shares manually using Lido's
+        // `getSharesByPooledEth` function.
+        else if (
+            block.chainid == 1 && address(_hyperdrive) == LEGACY_SDAI_HYPERDRIVE
```

Comment:
I didn't see a specific test for this case - might be a good idea

Reply:
Yeah, good call.